### PR TITLE
Enumerable methods

### PR DESCRIPTION
**Pull request by santiago-rodrig**
### Dear reviewer

The following tasks have been completed.

 - [x] `my_each`
 - [x] `my_each_with_index`
 - [x] `my_select`
 - [x] `my_all?`
 - [x] `my_any?`
 - [x] `my_none?`
 - [x] `my_count`
 - [x] `my_map`
 - [x] `my_inject`

If anything is wrong or misplaced please let me know and I'll fix it. Thanks for your time.

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    if block_given?
+      my_each { |k, v| return false if yield k, v } if is_a? Hash
+      my_each { |v| return false if yield v } unless is_a? Hash
+    else
+      return false unless is_a? Array
+
+      my_each { |v| return false if v } if is_a? Array
+    end
+    true
+  end
```
Comment (maelfosso) [inline enumerable.rb]:
```ruby
irb(main):013:0> array
=> [0, 4, 8, 2, 2, 3, 5, 7, 0, 1, 5, 4, 4, 2, 1, 8, 1, 6, 5, 3, 3, 3, 5, 8, 6, 5, 2, 3, 5, 2, 2, 3, 7, 0, 5, 5, 5, 3, 8, 0, 6, 5, 2, 0, 3, 5, 2, 4, 4, 0, 1, 6, 8, 7, 1, 5, 5, 3, 7, 3, 8, 1, 4, 3, 2, 8, 4, 0, 3, 8, 7, 0, 2, 3, 8, 4, 4, 4, 7, 6, 3, 3, 5, 4, 1, 8, 2, 4, 4, 2, 5, 4, 4, 3, 7, 2, 0, 6, 2, 1]
irb(main):014:0> array[0] = 'hi'
=> "hi"
irb(main):015:0> array.none?(String)
=> false
irb(main):016:0> array.my_none?(String)
=> true
irb(main):017:0> 
```

Comment (maelfosso) [reply enumerable.rb]:
```ruby
irb(main):022:0> words = %w[dog door rod blade]
=> ["dog", "door", "rod", "blade"]
irb(main):023:0> words.none?(/d/)
=> false
irb(main):024:0> words.my_none?(/d/)
=> true
irb(main):025:0> 
```

Comment (maelfosso) [reply enumerable.rb]:
```ruby
irb(main):027:0> words
=> ["dog", "door", "rod", "blade"]
irb(main):028:0> words[0] = 5
=> 5
irb(main):029:0> words.none?(5)
=> false
irb(main):030:0> words.my_none?(5)
=> true
irb(main):031:0> 
```

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+      return counter if is_a? Hash
+
+      my_each { |v| counter += 1 if v == arg }
+    end
+    if args.positive?
+      my_each { |v| counter += 1 if yield v } if block_given?
+      return size unless block_given?
+    end
+    counter
+  end
```
Comment (maelfosso) [inline enumerable.rb]:
```ruby
irb(main):035:0> array
=> [4, 8, 2, 2, 3, 5, 7, 0, 1, 5, 4, 4, 2, 1, 8, 1, 6, 5, 3, 3, 3, 5, 8, 6, 5, 2, 3, 5, 2, 2, 3, 7, 0, 5, 5, 5, 3, 8, 0, 6, 5, 2, 0, 3, 5, 2, 4, 4, 0, 1, 6, 8, 7, 1, 5, 5, 3, 7, 3, 8, 1, 4, 3, 2, 8, 4, 0, 3, 8, 7, 0, 2, 3, 8, 4, 4, 4, 7, 6, 3, 3, 5, 4, 1, 8, 2, 4, 4, 2, 5, 4, 4, 3, 7, 2, 0, 6, 2, 1]
irb(main):036:0> array.count
=> 99
irb(main):037:0> array.my_count
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):37
        1: from enumerable.rb:164:in `my_count'
NoMethodError (undefined method `positive?' for []:Array)
irb(main):038:0> 
```

Comment (maelfosso) [reply enumerable.rb]:
```ruby
irb(main):040:0> array
=> [4, 8, 2, 2, 3, 5, 7, 0, 1, 5, 4, 4, 2, 1, 8, 1, 6, 5, 3, 3, 3, 5, 8, 6, 5, 2, 3, 5, 2, 2, 3, 7, 0, 5, 5, 5, 3, 8, 0, 6, 5, 2, 0, 3, 5, 2, 4, 4, 0, 1, 6, 8, 7, 1, 5, 5, 3, 7, 3, 8, 1, 4, 3, 2, 8, 4, 0, 3, 8, 7, 0, 2, 3, 8, 4, 4, 4, 7, 6, 3, 3, 5, 4, 1, 8, 2, 4, 4, 2, 5, 4, 4, 3, 7, 2, 0, 6, 2, 1]
irb(main):041:0> array.count(3)
=> 16
irb(main):042:0> array.my_count(3)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):42
        1: from enumerable.rb:164:in `my_count'
NoMethodError (undefined method `positive?' for [3]:Array)
irb(main):043:0> 
```

Comment (maelfosso) [reply enumerable.rb]:
```ruby
irb(main):047:0> block = proc { |num| num < 9 / 2 }
=> #<Proc:0x0000559dc17185b8@(irb):47>
irb(main):048:0> array.count(&block)
=> 46
irb(main):049:0> array.my_count(&block)
Traceback (most recent call last):
        3: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        2: from (irb):49
        1: from enumerable.rb:164:in `my_count'
NoMethodError (undefined method `positive?' for []:Array)
irb(main):050:0> 
```

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    if args.empty?
+      my_each { |k, v| mapped << yield(k, v) } if is_a? Hash
+      my_each { |v| mapped << yield(v) } unless is_a? Hash
+    else
+      procedure = args[0]
+      my_each { |k, v| mapped << procedure.call(k, v) } if is_a? Hash
+      my_each { |v| mapped << procedure.call(v) } unless is_a? Hash
+    end
+    mapped
+  end
```
Comment (maelfosso) [inline enumerable.rb]:
```ruby
irb(main):052:0> array
=> [4, 8, 2, 2, 3, 5, 7, 0, 1, 5, 4, 4, 2, 1, 8, 1, 6, 5, 3, 3, 3, 5, 8, 6, 5, 2, 3, 5, 2, 2, 3, 7, 0, 5, 5, 5, 3, 8, 0, 6, 5, 2, 0, 3, 5, 2, 4, 4, 0, 1, 6, 8, 7, 1, 5, 5, 3, 7, 3, 8, 1, 4, 3, 2, 8, 4, 0, 3, 8, 7, 0, 2, 3, 8, 4, 4, 4, 7, 6, 3, 3, 5, 4, 1, 8, 2, 4, 4, 2, 5, 4, 4, 3, 7, 2, 0, 6, 2, 1]
irb(main):053:0> array.map
=> #<Enumerator: [4, 8, 2, 2, 3, 5, 7, 0, 1, 5, 4, 4, 2, 1, 8, 1, 6, 5, 3, 3, 3, 5, 8, 6, 5, 2, 3, 5, 2, 2, 3, 7, 0, 5, 5, 5, 3, 8, 0, 6, 5, 2, 0, 3, 5, 2, 4, 4, 0, 1, 6, 8, 7, 1, 5, 5, 3, 7, 3, 8, 1, 4, 3, 2, 8, 4, 0, 3, 8, 7, 0, 2, 3, 8, 4, 4, 4, 7, 6, 3, 3, 5, 4, 1, 8, 2, 4, 4, 2, 5, 4, 4, 3, 7, 2, 0, 6, 2, 1]:map>
irb(main):054:0> array.my_map
Traceback (most recent call last):
        7: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        6: from (irb):54
        5: from enumerable.rb:176:in `my_map'
        4: from enumerable.rb:67:in `my_each'
        3: from enumerable.rb:67:in `times'
        2: from enumerable.rb:67:in `block in my_each'
        1: from enumerable.rb:176:in `block in my_map'
LocalJumpError (no block given (yield))
irb(main):055:0>
```

**Comment by santiago-rodrig** [comment]
I did all the changes required, hope it passes now. **Thanks** for taking the time to review my code!

**Pull request merged**